src/java.base/share/classes/java/lang/reflect/code/op/ExtendedOp.java

```diff
@@ -3152,40 +3152,109 @@ static Block.Builder lowerTypePattern(Block.Builder endNoMatchBlock, Block.Build
-                    if (SHORT.equals(st) || CHAR.equals(st)) {
-                        st = INT;
+                Op p = null; // op that perform type check
+                Op c = null; // op that perform conversion
+                TypeElement s = target.type();
+                TypeElement t = targetType;
+                if (t instanceof PrimitiveType pt) {
+                    if (s instanceof ClassType cs) {
+                        // unboxing conversions
+                        ClassType box;
+                        if (cs.unbox().isEmpty()) { // s not a boxed type
+                            // e.g. Number -> int, narrowing + unboxing
+                            box = pt.box().orElseThrow();
+                            p = CoreOp.instanceOf(box, target);
+                        } else {
+                            // e.g. Float -> float, unboxing
+                            // e.g. Integer -> long, unboxing + widening
+                            box = cs;
+                        }
+                        c = invoke(MethodRef.method(box, t + "Value", t), target);
+                    } else {
+                        // primitive to primitive conversion
+                        PrimitiveType ps = ((PrimitiveType) s);
+                        if (isNarrowingPrimitiveConv(ps, pt) || isWideningPrimitiveConvWithCheck(ps, pt)
+                                || isWideningAndNarrowingPrimitiveConv(ps, pt)) {
+                            // e.g. int -> byte, narrowing
+                            // e,g. int -> float, widening with check
+                            // e.g. byte -> char, widening and narrowing
+                            MethodRef mref = convMethodRef(s, t);
+                            p = invoke(mref, target);
+                        }
+                        c = CoreOp.conv(targetType, target);
                     }
-                    String s = capitalize(st.toString());
-                    String t = capitalize(tt.toString());
-                    String mn = "is%sTo%sExact".formatted(s, t);
-                    JavaType exactConversionSupport = JavaType.type(ClassDesc.of("java.lang.runtime.ExactConversionsSupport"));
-                    MethodRef mref = MethodRef.method(exactConversionSupport, mn, BOOLEAN, st);
-                    p = currentBlock.op(invoke(mref, target));
-                } else {
-                    p = currentBlock.op(CoreOp.instanceOf(targetType, target));
+                } else if (s instanceof PrimitiveType ps) {
+                    // boxing conversions
+                    ClassType ct = (ClassType) t;
+                    ClassType box;
+                    if (ct.unbox().isEmpty()) {
+                        // e.g. int -> Number, boxing + widening
+                        box = ps.box().orElseThrow();
+                    } else {
+                        // e.g. byte -> Byte, boxing
+                        box = ct;
+                    }
```

Comment:
Can you just replace with `ClassType box = ps.box().orElseThrow()` ? (keep the examples).